packages/xo-server/src/xapi-stats.js

```diff
@@ -347,103 +308,91 @@ export default class XapiStats {
       )
     }
 
+    const currentTimeStamp = await getServerTimestamp(xapi, host.$ref)
+
     // Limit the number of http requests
-    const hostUuid = host.uuid
-
-    if (
-      !(
-        vmUuid !== undefined &&
-        get(this._statsByObject, [vmUuid, step]) === undefined
-      ) &&
-      get(this._statsByObject, [hostUuid, step, 'localTimestamp']) + step >
-        getCurrentTimestamp()
-    ) {
-      return this._getStats(hostUuid, step, vmUuid)
+    const stats = this._getCachedStats(uuid, step, currentTimeStamp)
+    if (stats !== undefined) {
+      return stats
     }
 
-    const timestamp = await this._getNextTimestamp(xapi, host, step)
-    const json = await this._getJson(xapi, host, timestamp, step)
+    const maxDuration = step * RRD_POINTS_PER_STEP[step]
+
+    // To avoid exceed the maxDuration, we add a step
+    const optimumTimestamp = currentTimeStamp - maxDuration + step
+    const json = await this._getJson(xapi, host, optimumTimestamp, step)
     if (json.meta.step !== step) {
       throw new FaultyGranularity(
         `Unable to get the true granularity: ${json.meta.step}`
       )
     }
```

Comment:
It's a bit sad to throw away valid fetched data…

Reply:
What we should do in this case ? updating the cache before throwing this error ?

Reply:
I think it would be better, don't you?